### PR TITLE
Lidar turn pid

### PR DESCRIPTION
**Pull request by NikhilSuresh24**
lidar pid makes lidar turn at about 240k mRPM +- 5k .

### CODE REVIEW

**Comment by c0d3rman** [comment]
Please update this PR's labels if it is tested @NikhilSuresh24 @carturn

**Comment by carturn** [comment]
Note that this has dependencies on [standard pull request #151](https://github.com/RoboticsTeam4904/standard/pull/151).

File: .classpath [hunk 1]
```diff
@@ -6,5 +6,7 @@
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/navx_frc.jar" sourcepath="USERLIBS_DIR/navx_frc-sources.jar"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/CTRLib.jar"/>
+	<classpathentry kind="var" path="opencv" sourcepath="opencv.sources"/>
+	<classpathentry kind="var" path="cscore" sourcepath="cscore.sources"/>
```
Comment (ajnadel) [inline .classpath]: @c0d3rman what should we do about this? If Eclipse adds these every time, let's just let it get merged in.

File: src/org/usfirst/frc4904/robot/Robot.java [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
+import org.usfirst.frc4904.robot.subsystems.Lidar;
 import org.usfirst.frc4904.standard.CommandRobotBase;
 import org.usfirst.frc4904.standard.commands.Noop;
+import edu.wpi.first.wpilibj.command.Command;
 import edu.wpi.first.wpilibj.command.Subsystem;
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
 
 public class Robot extends CommandRobotBase {
 	RobotMap map = new RobotMap();
+	Command LidarCommand;
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/Robot.java]: Fields should be camel case, but the leading letter should be lower case.

File: src/org/usfirst/frc4904/robot/Robot.java [hunk 3]
```diff
@@ -20,12 +24,21 @@ public void initialize() {
 		// Configure operator command chooser
 		operatorChooser.addDefault(new DefaultOperator());
 		// Initialize SmartDashboard display values
-		// SmartDashboard.putNumber(SmartDashboardKey.EXAMPLE.key, 0);
+		SmartDashboard.putNumber("LIDAR_P", Lidar.LIDAR_TURN_P);
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/Robot.java]: Check out `SmartDashboardKey` for the first argument.

File: src/org/usfirst/frc4904/robot/Robot.java [hunk 4]
```diff
@@ -20,12 +24,21 @@ public void initialize() {
...
 	@Override
 	public void teleopInitialize() {
 		teleopCommand = new Noop();
+		LidarCommand = new LidarTurner();
+		LidarCommand.start();
+		RobotMap.Component.lidarMC.reset();
+		Lidar.LIDAR_TURN_P = SmartDashboard.getNumber("LIDAR_P", Lidar.LIDAR_TURN_P);
+		Lidar.LIDAR_TURN_I = SmartDashboard.getNumber("LIDAR_I", Lidar.LIDAR_TURN_I);
+		Lidar.LIDAR_TURN_D = SmartDashboard.getNumber("LIDAR_D", Lidar.LIDAR_TURN_D);
+		RobotMap.Component.lidarMC.setPID(Lidar.LIDAR_TURN_P, Lidar.LIDAR_TURN_I, Lidar.LIDAR_TURN_D);
```
Comment (carturn) [inline src/org/usfirst/frc4904/robot/Robot.java]: In my opinion, we should remove the SmartDashboard code now that testing has finished.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 5]
```diff
@@ -51,20 +53,24 @@
...
 			public static final int elevatorEncoder = 0x606;
+			public static final int lidarTurnEncoder = 0x607;
 		}
 		
 		public static class CANMotor {}
 		
 		public static class PCM {}
 		
 		public static class Pneumatics {}
+		public static int lidarTurnP = 0;// put these in the constants section of RobotMap
```
Comment (leijurv) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: this comment...

File: src/org/usfirst/frc4904/robot/commands/LidarTurner.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+		requires(RobotMap.Component.lidar);
+	}
+
+	@Override
+	protected void initialize() {
+		try {
+			RobotMap.Component.lidar.reset();
+		}
+		catch (InvalidSensorException e) {
+			LogKitten.e("LIDAR error!");
```
Comment (NikhilSuresh24) [inline src/org/usfirst/frc4904/robot/commands/LidarTurner.java]: .

File: src/org/usfirst/frc4904/robot/commands/LidarTurner.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+			LogKitten.e(e);
+			LogKitten.e(e.getStackTrace());
+		}
+		RobotMap.Component.lidar.enableMotionController();
+		RobotMap.Component.lidar.set(240000);
```
Comment (NikhilSuresh24) [inline src/org/usfirst/frc4904/robot/commands/LidarTurner.java]: .

File: src/org/usfirst/frc4904/robot/subsystems/CANLidar.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	 * @return
+	 * 		The integer the Teensy returned for that mode
+	 *
+	 * @throws InvalidSensorException
+	 *         If the available data is more than one second old,
+	 *         this function will throw an InvalidSensorException
+	 *         to indicate that.
+	 * 
+	 */
+	public int[] readlidar() throws InvalidSensorException {
```
Comment (NikhilSuresh24) [inline src/org/usfirst/frc4904/robot/subsystems/CANLidar.java]: .

File: src/org/usfirst/frc4904/robot/subsystems/CANLidar.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+		write(new byte[] {0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01}); // Write to trigger read
```
Comment (NikhilSuresh24) [inline src/org/usfirst/frc4904/robot/subsystems/CANLidar.java]: .

File: src/org/usfirst/frc4904/robot/subsystems/LidarTurnPID.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package org.usfirst.frc4904.robot.subsystems;
```
Comment (NikhilSuresh24) [inline src/org/usfirst/frc4904/robot/subsystems/LidarTurnPID.java]: ?

**Review by ajnadel — COMMENTED** [review]
Dependencies resolved.

**Review by c0d3rman — APPROVED** [review]

**Review by c0d3rman — CHANGES_REQUESTED** [review]
Whoops, wrong PR.

**Comment by NikhilSuresh24** [comment]
Testerino'ed again.

**Review by c0d3rman — APPROVED** [review]

**Comment by ajnadel** [comment]
Service was good, "food tasted like bed bugs" but would probably return on a rainy Sunday night.

**Review by ajnadel — APPROVED** [review]
Service was good, "food tasted like bed bugs" but would probably return on a rainy Sunday night.

**Pull request merged**